### ISSUE TITLE
Allow ignoring debugger breaks when running in editor

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#6781]
### Describe the project you are working on

Some complex projects that sometimes have errors.

### Describe the problem or limitation you are having in your project

Sometimes I'm in the middle of doing something and it's unfinished, but I need to run another part of the project. Unfortunately the unfinished part spams hard errors (i.e. they stop the execution and I need to continue), which makes it impossible to play normally, because I have to continue every second or every frame. The error doesn't affect gameplay, it's the debugger being problematic.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be helpful if there was a way to ignore debugger breaks. If there is an error, don't stop, but continue normally. We already have Ignore Breakpoints option, this is just step further. It would allow to test project without being bothered by errors that don't affect the specific area you are testing right now.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I think it could be a new option, called "Ignore Debug Breaks", placed somewhere here:
![image](https://user-images.githubusercontent.com/2223172/235374973-aee9ac65-f4cf-42b4-954f-be242e693983.png)
(e.g. next to ignore breakpoints button)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Well, running the project outside editor counts as workaround, but I often test scenes using a custom running plugin, which can't be used outside the editor.

### Is there a reason why this should be core and not an add-on in the asset library?

It requires modifying editor functionality.

**Comment by dreed-sd** [comment]
Big +1 to this. There are multiple scenarios where this hurts. For example, we're having a team playtest with lots of non-technical folks running the game from editor. A bug is encountered where a sprite is not provided to some piece of code that is expecting it, and thus hits an error about dereferencing a null. This will not affect the gameplay though, so can be safely ignored. But everyone gets this jarring experience of having the editor take focus -- they say the game crashed, because that what it seems like to them. You inform them it isn't a crash, but it doesn't matter, they'll keep calling it one because that's what the experience is like. You tell them to hit the little 'keep going' button, and then they also have to know to click on the game window again to bring it back to the foreground. And depending on the issue, this might happen very frequently during the session. It completely wrecks your workflow entirely.

I appreciate being able to break on an error, that's a required feature for a good debugger. But I have never work with a debugger that has 'break on error' the default behavior, and certainly never worked with one that didn't even have the option to turn it off!

In a real build errors just cause whatever to be logged and execution continues. Shouldn't we be able to replicate what that experience is in editor?

I think 'break on error' should be optional, and it should be off by default.

**Comment by KoBeWi** [comment]
> In a real build errors just cause whatever to be logged and execution continues. Shouldn't we be able to replicate what that experience is in editor?

Note that you can run project directly, without opening editor (and without exporting). Just the Run option in Project Manager.

**Comment by dreed-sd** [comment]
> Note that you can run project directly, without opening editor (and without exporting). Just the Run option in Project Manager.

All this time I never knew that was a thing. Thanks! 

I still stand with my conviction though! :)

**Comment by Calinou** [comment]
> I think 'break on error' should be optional, and it should be off by default.

Having it off by default would mean developers would easily miss errors, especially if playing in fullscreen or with the editor window hidden somehow (which is common on single-monitor setups).

**Comment by dreed-sd** [comment]
> fail fast and ensure the error is fixed before it reaches production

I agree with that wholeheartedly, but there's always nuance. Here's another example. I have a suite of 1000s of tests written with GUT. A bug is introduced, which causes a runtime error while running the test suite. I want the running test to fail with error information so I can look into it. But instead, I hit the debugger, and I have no idea what test is even running without spelunking through the stack. Whatever the case, hitting resume is only the start of the adventure -- there's probably another error right after it, in the same test, since after all, it hit some unexpected condition already, there are probably more issues after it. Again and again. Not infrequently I have to spam the resume button. Sometimes there are so many errors to get through you just have to stop the test run. I'd much rather have all tests run and gather error information to examine afterward. If I need to dig deeper into a specific test, I'll debug it individually. Oftentimes, I won't even need to though, because whatever was logged when it failed was enough to know how to fix it. I just don't think forcing runs to be debug sessions is the right user experience.

**Comment by mihe** [comment]
> But I have never work with a debugger that has 'break on error' the default behavior, and certainly never worked with one that didn't even have the option to turn it off!

I suppose it's worth highlighting a key difference between GDScript and languages like C# and Java, and I suppose C++ as well to some degree, which is the notion of handled (caught) versus unhandled exceptions. Debuggers like Visual Studio are often configured by default to break on unhandled exceptions, but not handled exceptions, and since the unit testing frameworks do in fact handle/catch the exceptions thrown within unit tests you never end up with a debugger break when running them.

If you were to enable breaking on handled exceptions in something like Visual Studio (which you rarely, if ever, do) you would find yourself with a similar behavior as what Godot has right now, where you're just hammering the resume button all the time.

Godot/GDScript obviously has no notion of exceptions, and (as far as I know) no notion of "handling" errors either, so you end up having everything under the same umbrella, meaning (assuming you don't add this notion) it might be better to err on the side of treating everything as an "unhandled error" rather than a "handled error" and break by default.

With that said, having the option that's proposed/implemented here seems like a no-brainer, given the precedence for being able to disable breaking on unhandled exceptions in a lot of other debuggers.
